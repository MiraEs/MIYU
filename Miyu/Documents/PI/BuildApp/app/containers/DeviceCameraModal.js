import React, {
	PureComponent,
} from 'react';
import PropTypes from 'prop-types';
import {
	CameraRoll,
	StyleSheet,
	View,
	TouchableOpacity,
} from 'react-native';
import { withScreen } from 'BuildLibrary';
import styles from '../lib/styles';
import Camera from 'react-native-camera';
import Icon from 'react-native-vector-icons/Ionicons';
import helpers from '../lib/helpers';
import { withNavigation } from '@expo/ex-navigation';

const componentStyles = StyleSheet.create({
	container: {
		flex: 1,
		justifyContent: 'center',
		alignItems: 'flex-end',
		backgroundColor: styles.colors.lightGray,
	},
	cameraControls: {
		position: 'absolute',
		bottom: 0,
		left: 0,
		right: 0,
		backgroundColor: styles.colors.underlayGrey,
		height: 68,
	},
	toggleFrontBack: {
		position: 'absolute',
		right: 10,
		top: 10,
	},
	takePicture: {
		alignSelf: 'center',
	},
	closeButton: {
		position: 'absolute',
		top: 15,
		left: 15,
	},
	camera: {
		position: 'absolute',
		top: 0,
		left: 0,
		right: 0,
		bottom: 0,
	},
	icon: {
		height: 65,
		width: 65,
	},
	iconSmall: {
		height: 50,
		width: 50,
	},
});

export class DeviceCameraModal extends PureComponent {
	state = {
		cameraType: Camera.constants.Type.back,
	};

	setScreenTrackingInformation() {
		return {
			name: 'build:app:devicecameramodal',
		};
	}

	toggleFrontBackCamera = () => {
		const { back, front } = Camera.constants.Type;
		this.setState((prevState) => ({
			cameraType: prevState.cameraType === back ? front : back,
		}));
	};

	handleImageTaken = ({ path = '' }) => {
		if (path) {
			if (helpers.isIOS()) {
				this.props.onImageCapture(path);
			} else {
				// on Android we do this extra step because we get a file:/// path back
				// and the app cant actually find that file for uploading. So instead we
				// save it as a temp file initially and then save it to the camera roll
				// manually before uploading...
				CameraRoll.saveToCameraRoll(path).then((path) => {
					this.props.onImageCapture(path);
				});
			}
		}
	};

	takePicture = () => {
		this.cam.capture().then(this.handleImageTaken);
	};

	closeCamera = () => {
		this.props.navigator.pop();
	};

	render() {
		const captureTarget = helpers.isIOS() ? Camera.constants.CaptureTarget.cameraRoll : Camera.constants.CaptureTarget.temp;
		return (
			<View style={componentStyles.container}>
				<Camera
					ref={(node) => this.cam = node}
					aspect={Camera.constants.Aspect.fit}
					captureTarget={captureTarget}
					keepAwake={true}
					style={componentStyles.camera}
					type={this.state.cameraType}
				/>
				<View style={componentStyles.cameraControls}>
					<TouchableOpacity
						onPress={this.closeCamera}
						style={componentStyles.closeButton}
					>
						<Icon
							name={helpers.getIcon('close')}
							size={50}
							color="white"
							style={componentStyles.iconSmall}
						/>
					</TouchableOpacity>
					<TouchableOpacity
						onPress={this.takePicture}
						style={componentStyles.takePicture}
					>
						<Icon
							name="ios-radio-button-on"
							size={65}
							color="white"
							style={componentStyles.icon}
						/>
					</TouchableOpacity>
					<TouchableOpacity
						onPress={this.toggleFrontBackCamera}
						style={componentStyles.toggleFrontBack}
					>
						<Icon
							name="ios-reverse-camera-outline"
							size={50}
							color="white"
							style={componentStyles.iconSmall}
						/>
					</TouchableOpacity>
				</View>
			</View>
		);
	}
}

DeviceCameraModal.propTypes = {
	navigator: PropTypes.shape({
		pop: PropTypes.func,
	}),
	onImageCapture: PropTypes.func.isRequired,
};

DeviceCameraModal.defaultProps = {
	onImageCapture: helpers.noop,
};

export default withNavigation(withScreen(DeviceCameraModal));
