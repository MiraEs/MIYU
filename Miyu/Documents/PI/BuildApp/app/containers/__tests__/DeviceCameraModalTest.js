jest.mock('react-native');
jest.mock('BuildNative');
jest.mock('BuildLibrary');
jest.mock('react-native-vector-icons/Ionicons', () => 'Icon');
jest.mock('react-native-camera');

import React from 'react';
import { DeviceCameraModal } from '../DeviceCameraModal';
import renderer from 'react-test-renderer';

const defaultProps = {
	onImageCapture: jest.fn(),
	navigator: {
		pop: jest.fn(),
	},
};

describe('DeviceCameraModal', () => {

	it('should render', () => {
		const tree = renderer.create(
			<DeviceCameraModal {...defaultProps} />
		);
		expect(tree.toJSON()).toMatchSnapshot();
	});
});

describe('DeviceCameraModal functions', () => {
	it('state', () => {
		const wrapper = renderer.create(
			<DeviceCameraModal {...defaultProps} />
		).getInstance();
		expect(wrapper.state).toMatchSnapshot();
	});

	it('setScreenTrackingInformation', () => {
		const result = renderer.create(
			<DeviceCameraModal {...defaultProps} />
		).getInstance().setScreenTrackingInformation();
		expect(result).toMatchSnapshot();
	});

	it('toggleFrontBackCamera', () => {
		const wrapper = renderer.create(
			<DeviceCameraModal {...defaultProps} />
		).getInstance();
		wrapper.toggleFrontBackCamera();
		expect(wrapper.state.cameraType).toEqual('front');
		wrapper.toggleFrontBackCamera();
		expect(wrapper.state.cameraType).toEqual('back');
	});

	it('handleImageTaken', () => {
		const wrapper = renderer.create(
			<DeviceCameraModal {...defaultProps} />
		).getInstance();
		wrapper.handleImageTaken({ path: 'path' });
		expect(defaultProps.onImageCapture).toHaveBeenCalledWith('path');
	});

	it('takePicture', () => {
		const wrapper = renderer.create(
			<DeviceCameraModal {...defaultProps} />
		).getInstance();
		wrapper.takePicture();
		expect(wrapper.cam.capture).toHaveBeenCalled();
	});

	it('closeCamera', () => {
		renderer.create(
			<DeviceCameraModal {...defaultProps} />
		).getInstance().closeCamera();
		expect(defaultProps.navigator.pop).toHaveBeenCalled();
	});

});
