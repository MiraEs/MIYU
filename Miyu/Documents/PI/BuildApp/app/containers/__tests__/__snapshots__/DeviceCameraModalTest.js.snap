// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviceCameraModal functions setScreenTrackingInformation 1`] = `
Object {
  "name": "build:app:devicecameramodal",
}
`;

exports[`DeviceCameraModal functions state 1`] = `
Object {
  "cameraType": "back",
}
`;

exports[`DeviceCameraModal should render 1`] = `
<View
  style={
    Object {
      "alignItems": "flex-end",
      "backgroundColor": "#F2F2F2",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Camera
    aspect={undefined}
    captureTarget={undefined}
    keepAwake={true}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    type="back"
  />
  <View
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.5)",
        "bottom": 0,
        "height": 68,
        "left": 0,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <TouchableOpacity
      onPress={[Function]}
      style={
        Object {
          "left": 15,
          "position": "absolute",
          "top": 15,
        }
      }
    >
      <Icon
        color="white"
        name="ios-close"
        size={50}
        style={
          Object {
            "height": 50,
            "width": 50,
          }
        }
      />
    </TouchableOpacity>
    <TouchableOpacity
      onPress={[Function]}
      style={
        Object {
          "alignSelf": "center",
        }
      }
    >
      <Icon
        color="white"
        name="ios-radio-button-on"
        size={65}
        style={
          Object {
            "height": 65,
            "width": 65,
          }
        }
      />
    </TouchableOpacity>
    <TouchableOpacity
      onPress={[Function]}
      style={
        Object {
          "position": "absolute",
          "right": 10,
          "top": 10,
        }
      }
    >
      <Icon
        color="white"
        name="ios-reverse-camera-outline"
        size={50}
        style={
          Object {
            "height": 50,
            "width": 50,
          }
        }
      />
    </TouchableOpacity>
  </View>
</View>
`;
